From dev-bounces@openvswitch.org Fri Oct  2 18:59:40 2009
Received: from SMTP.EU.CITRIX.COM (10.30.244.21) by
	LONPMAILMX02.citrite.net (10.30.224.163) with Microsoft SMTP Server id
	8.1.393.1; Fri, 2 Oct 2009 18:59:40 +0100
Received: from dime139.dizinc.com ([66.7.205.111])  by SMTP.EU.CITRIX.COM
	with ESMTP; 02 Oct 2009 17:59:39 +0000
Received: from localhost ([127.0.0.1]:37666 helo=dime139.dizinc.com)	by
	dime139.dizinc.com with esmtp (Exim 4.69)	(envelope-from
	<dev-bounces@openvswitch.org>)	id 1MtmPj-0005Yk-AN; Fri, 02 Oct 2009
	13:59:31 -0400
Received: from [209.172.104.10] (port=60246 helo=leadville.nicira.com)	by
	dime139.dizinc.com with esmtps (TLSv1:AES256-SHA:256) (Exim 4.69)
	(envelope-from <blp@nicira.com>) id 1MtmPe-0005Y9-Nh	for
	dev@openvswitch.org; Fri, 02 Oct 2009 13:59:27 -0400
Received: from hardrock.nicira.com ([172.16.0.20])	by leadville.nicira.com
	with esmtp (Exim 4.67)	(envelope-from <blp@nicira.com>)	id
	1MtmPO-0000w8-Nf; Fri, 02 Oct 2009 10:59:10 -0700
Received: from blp by hardrock.nicira.com with local (Exim 4.67)
	(envelope-from <blp@hardrock.nicira.com>)	id 1MtmPe-00056b-GV; Fri, 02 Oct
	2009 10:59:26 -0700
From: Ben Pfaff <blp@nicira.com>
To: "dev@openvswitch.org" <dev@openvswitch.org>
CC: Ben Pfaff <blp@nicira.com>
Sender: "dev-bounces@openvswitch.org" <dev-bounces@openvswitch.org>
Date: Fri, 2 Oct 2009 18:59:22 +0100
Subject: [ovs-dev] [PATCH 1/2] vswitch: Factor out detection of internal
	interfaces into a new function.
Thread-Topic: [ovs-dev] [PATCH 1/2] vswitch: Factor out detection of
	internal	interfaces into a new function.
Thread-Index: AcpDih10GP224iZCQAmujETR6qjeIA==
Message-ID: <1254506363-19569-1-git-send-email-blp@nicira.com>
List-Help: <mailto:dev-request@openvswitch.org?subject=help>
List-Subscribe:
	<http://openvswitch.org/mailman/listinfo/dev_openvswitch.org>,
	<mailto:dev-request@openvswitch.org?subject=subscribe>
List-Unsubscribe:
	<http://openvswitch.org/mailman/options/dev_openvswitch.org>,
	<mailto:dev-request@openvswitch.org?subject=unsubscribe>
Accept-Language: en-GB, en-US
Content-Language: en-US
X-MS-Exchange-Organization-AuthAs: Anonymous
X-MS-Exchange-Organization-AuthSource: LONPMAILMX02.citrite.net
X-MS-Has-Attach: 
X-Auto-Response-Suppress: All
X-MS-TNEF-Correlator: 
x-sbrs: 3.2
x-mesageid: 7390310
x-ironport-server: lonpip01.citrite.net
x-remote-ip: 66.7.205.111
x-policy: $ACCEPTED
x-ironport-anti-spam-filtered: true
x-ironport-anti-spam-result:
	Ak8DAH/cxUpCB81vhGdsb2JhbACadAEBAQoHBAgJE7xghCwE
x-ironport-av: E=Sophos;i="4.44,495,1249257600";    d="scan'208";a="7390310"
x-antiabuse: Sender Address Domain - openvswitch.org
x-source: 
x-source-args: 
x-source-dir: 
list-id: <dev_openvswitch.org.openvswitch.org>
errors-to: dev-bounces@openvswitch.org
list-post: <mailto:dev@openvswitch.org>
list-archive: <http://openvswitch.org/pipermail/dev_openvswitch.org>
x-beenthere: dev@openvswitch.org
x-mailman-version: 2.1.11.cp3
Content-Type: text/plain; charset="us-ascii"
MIME-Version: 1.0
X-Evolution-Source: imap://ianca@lonpmailmx01.citrite.net/
Content-Transfer-Encoding: 8bit

The following commit needs to use this same logic, so break it out into
a function to avoid redundancy.
---
 vswitchd/bridge.c |   41 ++++++++++++++++++++++++++++++-----------
 1 files changed, 30 insertions(+), 11 deletions(-)

diff --git a/vswitchd/bridge.c b/vswitchd/bridge.c
index b525f3d..2d788ae 100644
--- a/vswitchd/bridge.c
+++ b/vswitchd/bridge.c
@@ -242,6 +242,7 @@ static void iface_destroy(struct iface *);
 static struct iface *iface_lookup(const struct bridge *, const char *name);
 static struct iface *iface_from_dp_ifidx(const struct bridge *,
                                          uint16_t dp_ifidx);
+static bool iface_is_internal(const struct bridge *, const char *name);
 
 /* Hooks into ofproto processing. */
 static struct ofhooks bridge_ofhooks;
@@ -464,18 +465,8 @@ bridge_reconfigure(void)
             bool internal;
             int error;
 
-            /* It's an internal interface if it's marked that way, or if
-             * it's a bonded interface for which we're faking up a network
-             * device. */
-            internal = cfg_get_bool(0, "iface.%s.internal", if_name);
-            if (cfg_get_bool(0, "bonding.%s.fake-iface", if_name)) {
-                struct port *port = port_lookup(br, if_name);
-                if (port && port->n_ifaces > 1) {
-                    internal = true;
-                }
-            }
-
             /* Add to datapath. */
+            internal = iface_is_internal(br, if_name);
             error = dpif_port_add(br->dpif, if_name,
                                   internal ? ODP_PORT_INTERNAL : 0, NULL);
             if (error == EFBIG) {
@@ -3116,6 +3107,34 @@ iface_from_dp_ifidx(const struct bridge *br, uint16_t dp_ifidx)
 {
     return port_array_get(&br->ifaces, dp_ifidx);
 }
+
+/* Returns true if 'iface' is the name of an "internal" interface on bridge
+ * 'br', that is, an interface that is entirely simulated within the datapath.
+ * The local port (ODPP_LOCAL) is always an internal interface.  Other local
+ * interfaces are created by setting "iface.<iface>.internal = true".
+ *
+ * In addition, we have a kluge-y feature that creates an internal port with
+ * the name of a bonded port if "bonding.<bondname>.fake-iface = true" is set.
+ * This feature needs to go away in the long term.  Until then, this is one
+ * reason why this function takes a name instead of a struct iface: the fake
+ * interfaces created this way do not have a struct iface. */
+static bool
+iface_is_internal(const struct bridge *br, const char *iface)
+{
+    if (!strcmp(iface, br->name)
+        || cfg_get_bool(0, "iface.%s.internal", iface)) {
+        return true;
+    }
+
+    if (cfg_get_bool(0, "bonding.%s.fake-iface", iface)) {
+        struct port *port = port_lookup(br, iface);
+        if (port && port->n_ifaces > 1) {
+            return true;
+        }
+    }
+
+    return false;
+}
 
 /* Port mirroring. */
 
-- 
1.6.3.3


_______________________________________________
dev mailing list
dev@openvswitch.org
http://openvswitch.org/mailman/listinfo/dev_openvswitch.org
